codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -498,15 +543,74 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx,
+                paddingBytes(padding) <= 64) {
+                // no need for a continuation either!
+                try {
+                    int paddingBytes = paddingBytes(padding);
+                    flags.endOfHeaders(true);
+                    final int bufBytes;
+                    if (paddingBytes > 0) {
+                        bufBytes = HEADERS_FRAME_HEADER_LENGTH + fragmentBytes + paddingBytes;
+                    } else {
+                        bufBytes = HEADERS_FRAME_HEADER_LENGTH + fragmentBytes;
+                    }
+                    ByteBuf buf = ctx.alloc().buffer(bufBytes);
+                    int payloadLength = fragmentBytes + nonFragmentBytes;
+                    writeFrameHeaderInternal(buf, payloadLength, HEADERS, flags, streamId);
+                    writePaddingLength(buf, padding);
+                    if (hasPriority) {
+                        buf.writeInt(exclusive? (int) (0x80000000L | streamDependency) : streamDependency);
+                        // Adjust the weight so that it fits into a single byte on the wire.
+                        buf.writeByte(weight - 1);
+                    }
+                    buf.writeBytes(headerBlock, fragmentBytes);
+                    if (paddingBytes > 0) {
+                        // this is quite fast assuming small padding
+                        // see https://github.com/netty/netty/pull/13693
+                        buf.writeZero(paddingBytes);
+                    }
+                    ctx.write(buf, promise);
+                } catch (Throwable t) {
+                    promise.setFailure(t);
+                    PlatformDependent.throwException(t);
+                }
+                return promise;
+            } else {
+                return writeHeadersPartsInternal(ctx, streamId, padding, hasPriority, streamDependency,
+                                                 weight, exclusive, promise, headerBlock, fragmentBytes,
+                                                 flags, nonFragmentBytes);
+            }
+        } finally {
+            if (headerBlock != null) {
+                headerBlock.release();
+            }
+        }
+    }
+
+    private ChannelPromise writeHeadersPartsInternal(ChannelHandlerContext ctx, int streamId, int padding,
+                                                     boolean hasPriority, int streamDependency, short weight,
+                                                     boolean exclusive, ChannelPromise promise, ByteBuf headerBlock,
+                                                     int fragmentBytes, Http2Flags flags, int nonFragmentBytes) {
+        final SimpleChannelPromiseAggregator promiseAggregator =
+                new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+        try {
+            // Read the first fragment (possibly everything).
+            ByteBuf fragment = headerBlock.readRetainedSlice(fragmentBytes);
```

Comment:
We should ensure we release this one in the case of an error otherwise we might leak

Reply:
I see that this was missing in the original code as well, so I could fix it separately or contextual to this change if you wish

Reply:
maybe do a PR for the this first and then let us merge it and rebase this one . WDYT ?

Reply:
Yep I think is a good idea, mixing up them will loose the fix, which is not nice :+1:

Reply:
And I was wrong, it is handled, but just once :( 
I see how to fix this in this PR at this point, given that I already have to better reorganize the code...